File: Composite/CompositePattern.java
import java.util.*;

interface Node {
    boolean isComposite();
    String getName();
    Node getNodeByName(String name);
    int addNode(Node node);
    int ls();
    void tree();
    Map<String, Node> getSubNode();
}

class File implements Node {
    private String name;

    public boolean isComposite() {
        return false;
    }

    public File(String name) {
        if (name == null)
            throw new RuntimeException("name == null");
        this.name = name;
    }

    public String getName() {
        return name;
    }
    
    public int addNode(Node node) {
        System.out.println(name + " is not a directory");
        return 1;
    }

    public Node getNodeByName(String name) {
        throw new RuntimeException(name + " is not a directory");
    }

    public Map<String, Node> getSubNode() {
        throw new RuntimeException(name + " is not a directory");
    }

    public int ls() {
        System.out.println(name + " is not a directory");
        return 1;
    }

    public void tree() {
        System.out.println(name);
    }
}

class Directory implements Node {
    private String name;
    private Map<String, Node> map = new HashMap<>();

    public boolean isComposite() {
        return true;
    }

    public Directory(String name) {
        if (name == null)
            throw new RuntimeException("name == null");
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public int addNode(Node node) {
        if (node == null) {
            System.out.println("node == null");
            return 2;
        }
        if (map.containsKey(node.getName())) {
            System.out.println(node.getName() + " already exists");
            return 3;
        }
        map.put(node.getName(), node);
        return 0;
    }

    public int ls() {
        for (Map.Entry<String, Node> entry: map.entrySet()) {
            System.out.println(entry.getKey());
        }
        return 0;
    }

    public Node getNodeByName(String name) {
        if (name == null || !map.containsKey(name))
            throw new RuntimeException("name not exists");
        return map.get(name);
    }

    public void tree() {
        tree("", this);
    }

    public static String repeat(String string, int count) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < count; ++i)
            sb.append(string);
        return sb.toString();
    }

    public Map<String, Node> getSubNode() {
        return map;
    }

    private static void tree(String prefix, Node node) {
        System.out.println(prefix + node.getName());
        if (!node.isComposite())
            return;
        for (Map.Entry<String, Node> entry: node.getSubNode().entrySet()) {
            String newPrefix = prefix + repeat(
                " ", 
                node.getName().length() + 1);
            tree(newPrefix, entry.getValue());
        }
    }
}

public class CompositePattern {
    public static void main(String[] args) {
        Node root = new Directory("/");
        root.addNode(new Directory("etc"));
        root.addNode(new Directory("lib"));
        root.getNodeByName("lib").addNode(new File("ld.so"));
        Node etc = root.getNodeByName("etc");
        etc.addNode(new File("my.cnf"));
        Node myCnf = etc.getNodeByName("my.cnf");
        myCnf.tree();
        root.tree();
    }
}

File: Composite/CompositePattern.py
from abc import ABCMeta, abstractmethod
import os


class Node(object):
    __metaclass__ = ABCMeta

    @abstractmethod
    def is_composite(self):
        pass

    @abstractmethod
    def get_name(self):
        pass

    @abstractmethod
    def get_sub_nodes(self):
        pass

    @abstractmethod
    def get_node_by_name(self, name):
        pass

    @abstractmethod
    def add_node(self, node):
        pass

    @abstractmethod
    def get_files(self):
        pass

    @abstractmethod
    def delete_node_by_name(self, name):
        pass

    @classmethod
    def tree(cls, node, prefix=""):
        if not node.is_composite():
            return [os.path.join(prefix, node.get_name())]

        nodes = []
        for sub_node in node.get_sub_nodes():
            if not sub_node.is_composite():
                nodes.append(os.path.join(prefix, node.get_name(), sub_node.get_name()))
                continue
            nodes.extend(cls.tree(sub_node, os.path.join(prefix, node.get_name())))
        return nodes


class File(Node):
    def __init__(self, name):
        self._name = name

    def is_composite(self):
        return False

    def get_name(self):
        return self._name

    def get_sub_nodes(self):
        raise RuntimeError()

    def get_node_by_name(self, name):
        raise RuntimeError()

    def add_node(self, node):
        raise RuntimeError()

    def get_files(self):
        return self._name

    def delete_node_by_name(self, name):
        raise RuntimeError()


class Directory(Node):
    def __init__(self, name):
        self._name = name
        self._map = {}

    def is_composite(self):
        return True

    def get_name(self):
        return self._name

    def get_sub_nodes(self):
        return self._map.values()

    def get_node_by_name(self, name):
        return self._map[name]

    def add_node(self, node):
        self._map[node.get_name()] = node

    def delete_node_by_name(self, name):
        self._map.pop(name)

    def get_files(self):
        return self.tree(self)


if __name__ == "__main__":
    root = Directory("")
    root.add_node(Directory("etc"))
    root.add_node(Directory("lib"))
    etc = root.get_node_by_name("etc")
    lib = root.get_node_by_name("lib")
    etc.add_node(File("my.cnf"))
    etc.add_node(File("ld.so.conf"))
    etc.add_node(Directory("yum.repos.d"))
    lib.add_node(File("ld.so"))
    yum_repos_d = etc.get_node_by_name("yum.repos.d")
    yum_repos_d.add_node(File("CentOS-Base.repo"))

    print(root.get_files())
